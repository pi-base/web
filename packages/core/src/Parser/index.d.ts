declare module 'rehype-truncate' {
  import { Root } from 'mdast'
  import { Transformer } from 'unified'

  type Opts = {
    disable?: boolean
    ellipses?: string
    maxChars: number
    ignoreTags?: string[]
  }

  export default function rehypeTruncate(opts: Opts): Transformer<Root, Root>
}
