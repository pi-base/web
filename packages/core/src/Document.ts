import * as z from 'zod'
import * as yaml from 'js-yaml'

// TODO: unify with logic in packages/compiler
export function parseDocument<T>(schema: z.ZodSchema<T>, contents: string) {
  const match = contents.match(
    /^(---)?\s*(?<frontmatter>[\s\S]*?)\s*---(?<body>[\s\S]*)/,
  )
  if (!match?.groups) {
    return
  }

  const { frontmatter, body } = match.groups
  const meta = yaml.load(frontmatter)
  const data = { description: body.trim() }
  const raw = typeof meta === 'object' ? { ...meta, ...data } : data

  return schema.safeParse(raw)
}
