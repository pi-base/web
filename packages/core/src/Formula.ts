import { z } from 'zod'
import { parse as _parse } from './Formula/Grammar.js'

import { union } from './Util.js'

type F<P> =
  | { kind: 'atom'; property: P; value: boolean }
  | { kind: 'or'; subs: F<P>[] }
  | { kind: 'and'; subs: F<P>[] }

function atomSchema<P>(p: z.ZodSchema<P>): z.ZodSchema<F<P>> {
  return z.object({
    kind: z.literal('atom'),
    property: p,
    value: z.boolean(),
  }) as any
}

const andSchema = <P>(p: z.ZodSchema<P>) =>
  z.object({
    kind: z.literal('and'),
    subs: z.array(z.lazy(() => formulaSchema(p))),
  })

const orSchema = <P>(p: z.ZodSchema<P>) =>
  z.object({
    kind: z.literal('or'),
    subs: z.array(z.lazy(() => formulaSchema(p))),
  })

export const formulaSchema = <P>(p: z.ZodSchema<P>): z.ZodSchema<F<P>> =>
  z.union([atomSchema(p), andSchema(p), orSchema(p)])

export interface Atom<P> {
  kind: 'atom'
  property: P
  value: boolean
}

export interface And<P> {
  kind: 'and'
  subs: Formula<P>[]
}

export interface Or<P> {
  kind: 'or'
  subs: Formula<P>[]
}

export type Formula<P> = And<P> | Or<P> | Atom<P>

export function and<P>(...subs: Formula<P>[]): And<P> {
  return { kind: 'and', subs: subs }
}

export function or<P>(...subs: Formula<P>[]): Or<P> {
  return { kind: 'or', subs: subs }
}

export function atom<P>(p: P, v = true): Atom<P> {
  return { kind: 'atom', property: p, value: v }
}

export function properties<P>(f: Formula<P>): Set<P> {
  switch (f.kind) {
    case 'atom':
      return new Set([f.property])
    default:
      return union(...f.subs.map(properties))
  }
}

export function render<T>(f: Formula<T>, term: (t: T) => string): string {
  switch (f.kind) {
    case 'atom':
      // eslint-disable-next-line no-case-declarations
      const name = term(f.property)
      return f.value ? name : '¬' + name
    case 'and':
      return '(' + f.subs.map(sf => render(sf, term)).join(' ∧ ') + ')'
    case 'or':
      return '(' + f.subs.map(sf => render(sf, term)).join(' ∨ ') + ')'
  }
}

export function negate<P>(formula: Formula<P>): Formula<P> {
  switch (formula.kind) {
    case 'atom':
      return atom(formula.property, !formula.value)
    case 'and':
      return or(...formula.subs.map(negate))
    case 'or':
      return and(...formula.subs.map(negate))
  }
}

export function map<P, Q>(
  func: (p: Atom<P>) => Atom<Q>,
  formula: Formula<P>,
): Formula<Q> {
  switch (formula.kind) {
    case 'atom':
      return func(formula)
    default:
      return {
        ...formula,
        subs: formula.subs.map(sub => map(func, sub)),
      }
  }
}

export function mapProperty<P, Q>(
  func: (p: P) => Q,
  formula: Formula<P>,
): Formula<Q> {
  function mapAtom(a: Atom<P>): Atom<Q> {
    return { ...a, property: func(a.property) }
  }
  return map<P, Q>(mapAtom, formula)
}

export function compact<P>(f: Formula<P | undefined>): Formula<P> | undefined {
  return properties(f).has(undefined) ? undefined : (f as Formula<P>)
}

export function evaluate<T>(
  f: Formula<T>,
  traits: Map<T, boolean>,
): boolean | undefined {
  let result: boolean | undefined

  switch (f.kind) {
    case 'atom':
      if (traits.has(f.property)) {
        return traits.get(f.property) === f.value
      }
      return undefined
    case 'and':
      result = true // by default
      f.subs.forEach(sub => {
        if (result === false) {
          return
        }
        const sv = evaluate(sub, traits)
        if (sv === false) {
          // definitely false
          result = false
        } else if (result && sv === undefined) {
          // maybe false
          result = undefined
        }
      })
      return result
    case 'or':
      result = false
      f.subs.forEach(sub => {
        if (result === true) {
          return
        }
        const sv = evaluate(sub, traits)
        if (sv === true) {
          // definitely true
          result = true
        } else if (result === false && sv === undefined) {
          // maybe true
          result = undefined
        }
      })
      return result
  }
}

export function parse(q?: string): Formula<string> | undefined {
  if (!q) {
    return
  }

  let parsed
  try {
    // eslint-disable-next-line
    parsed = _parse(q)
  } catch {
    if (q && q.startsWith('(')) {
      return
    } else {
      return parse('(' + q + ')')
    }
  }

  return fromJSON(parsed as any)
}

type Serialized =
  | { and: Serialized[] }
  | { or: Serialized[] }
  | { property: string; value: boolean }
  | Record<string, boolean>

export function fromJSON(json: Serialized): Formula<string> {
  if ('and' in json && typeof json.and === 'object') {
    return and<string>(...json.and.map(fromJSON))
  } else if ('or' in json && typeof json.or === 'object') {
    return or<string>(...json.or.map(fromJSON))
  } else if ('property' in json && typeof json.property === 'string') {
    return atom<string>(json.property, json.value)
  }

  const entries = Object.entries(json)
  if (entries.length !== 1) {
    throw `cannot cast object with ${entries.length} keys to atom`
  }

  if (typeof entries[0][1] !== 'boolean') {
    throw `cannot cast object with non-boolean value`
  }

  return atom<string>(...entries[0])
}

export function toJSON(f: Formula<string>): Serialized {
  switch (f.kind) {
    case 'atom':
      return { [f.property]: f.value }
    case 'and':
      return { and: f.subs.map(toJSON) }
    case 'or':
      return { or: f.subs.map(toJSON) }
  }
}
