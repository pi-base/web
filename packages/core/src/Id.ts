export type Id = string
export type TraitId = { space: string; property: string }

export type Tagged =
  | SpaceId
  | PropertyId
  | TheoremId
  | { kind: 'trait'; space: number; property: number }

export type SpaceId = { kind: 'space'; id: number }
export type PropertyId = { kind: 'property'; id: number }
export type TheoremId = { kind: 'theorem'; id: number }

const pattern = /^(?<prefix>[spti])0*(?<id>\d+)/i

export function traitId({ space, property }: TraitId): string {
  return `${space}|${property}`
}

export function format(prefix: string, number: number): string {
  return `${prefix}${number.toString().padStart(6, '0')}`
}

export function trim(id: string): string {
  const match = pattern.exec(id)
  return match?.groups ? match.groups.id : id
}

export function tag(input: string): Tagged | null {
  const match = pattern.exec(input)
  if (!match || !match.groups) {
    return null
  }

  const id = parseInt(match.groups.id)
  switch (match.groups.prefix.toUpperCase()) {
    case 'S':
      return { kind: 'space', id }
    case 'P':
      return { kind: 'property', id }
    case 'T':
    case 'I':
      return { kind: 'theorem', id }
    default:
      return null
  }
}

export function toInt(id: string): number {
  const tagged = tag(id)
  if (tagged === null || tagged.kind === 'trait') {
    return 0
  } // TODO: return undefined

  return tagged.id
}

// TODO: these were extracted from other parallel-but-divergent implementations
// and should be unified.

type Pad = '' | '0' | '00' | '000' | '0000' | '00000' | '00000'
type XId<Prefix extends string> = `${Prefix}${Pad}${number}`

export type SId = XId<'S'>
export type PId = XId<'P'>
export type TId = XId<'T'>
export type SPId = [SId, PId]
export type EntityId = SId | PId | TId | SPId

export function isSpaceId(token: string): token is SId {
  return token.match(/^S\d{1,6}$/) !== null
}

export function isPropertyId(token: string): token is PId {
  return token.match(/^P\d{1,6}$/) !== null
}

export function isTheoremId(token: string): token is SId {
  return token.match(/^T\d{1,6}$/) !== null
}

export function isTraitId(pair: [string, string]): pair is SPId {
  return isSpaceId(pair[0]) && isPropertyId(pair[1])
}

export const idExp = /[PST]\d{1,6}/g

export function normalizeId(id: SId): SId
export function normalizeId(id: PId): PId
export function normalizeId(id: string) {
  return `${id[0]}${id.slice(1).padStart(6, '0')}`
}
