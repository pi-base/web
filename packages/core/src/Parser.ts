import { Data, Node } from 'unist'
import { Processor, unified } from 'unified'
import remarkParse from 'remark-parse'
import remarkMath from 'remark-math'
import remarkRehype from 'remark-rehype'
import rehypeKatex from 'rehype-katex'
import rehypeStringify from 'rehype-stringify'

import { links } from './Parser/links.js'
import { unnest } from './Parser/unnest.js'
import { truncation } from './Parser/truncation.js'

function noOp() {
  return (tree: Node) => tree
}

export type Options = {
  link?: boolean
  truncate?: boolean
}

function trace() {
  return (tree: Node) => {
    console.log(tree)
    return tree
  }
}

/**
 * Full parser chain translating Markdown to HTML, including rendering math and
 * custom Ï€-base syntax extensions.
 */
export function parser({
  link = true,
  truncate = false,
}: Partial<Options> = {}): Processor<Node<Data>, Node<Data>, Node<Data>, void> {
  return unified()
    .use(remarkParse)
    .use(link ? links : noOp)
    .use(unnest)
    .use(remarkMath)
    .use(truncate ? truncation : noOp)
    .use(remarkRehype)
    .use(rehypeKatex)
    .use(rehypeStringify)
}
