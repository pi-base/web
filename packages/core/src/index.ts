import { formulaSchema } from './Formula.js'
import { propertySchema } from './Property.js'
import { spaceSchema } from './Space.js'
import { theoremSchema } from './Theorem.js'
import { traitSchema } from './Trait.js'

export { type Bundle } from './Bundle.js'
export { type Formula, type And, type Atom, type Or } from './Formula.js'
export { parser } from './Parser.js'
export { type Property } from './Property.js'
export {
  ImplicationIndex,
  deduceTraits,
  disproveFormula,
  proveTheorem,
} from './Logic/index.js'
export { type Space } from './Space.js'
export { type Theorem } from './Theorem.js'
export { type Trait } from './Trait.js'
export { type Version } from './Bundle.js'

export * as bundle from './Bundle.js'
export * as formula from './Formula.js'
export * as Id from './Id.js'
export * as Ref from './Ref.js'

export * as TestUtils from './testUtils.js'

export const schemas = {
  formula: formulaSchema,
  property: propertySchema,
  space: spaceSchema,
  theorem: theoremSchema,
  trait: traitSchema
}