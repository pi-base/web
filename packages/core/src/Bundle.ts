/// <reference types="vite/client" />
import { z } from 'zod'
import { Id, traitId } from './Id.js'
import { Property, propertySchema } from './Property.js'
import { Space, spaceSchema } from './Space.js'
import { Theorem, theoremSchema } from './Theorem.js'
import { Trait, traitSchema } from './Trait.js'

export const defaultHost = import.meta.env?.VITE_PUBLIC_DATA_URL
  ? import.meta.env.VITE_PUBLIC_DATA_URL
  : import.meta.env?.DEV
  ? 'http://localhost:3141'
  : 'https://pi-base-bundles.s3.us-east-2.amazonaws.com'

export type Version = {
  ref: string
  sha: string
}

export type Bundle = {
  properties: Map<Id, Property>
  spaces: Map<Id, Space>
  traits: Map<Id, Trait<Id>>
  theorems: Map<Id, Theorem>
  version: Version
}

export const serializedSchema = z.object({
  properties: z.array(propertySchema),
  spaces: z.array(spaceSchema),
  theorems: z.array(theoremSchema),
  traits: z.array(traitSchema(z.string())),
  version: z.object({ ref: z.string(), sha: z.string() }),
})

export type Serialized = z.infer<typeof serializedSchema>

export function serialize(bundle: Bundle): Serialized {
  return serializedSchema.parse({
    properties: [...bundle.properties.values()],
    spaces: [...bundle.spaces.values()],
    theorems: [...bundle.theorems.values()],
    traits: [...bundle.traits.values()],
    version: bundle.version,
  })
}

export function deserialize(data: unknown): Bundle {
  const serialized = serializedSchema.parse(data)

  return {
    properties: indexBy(serialized.properties, p => p.uid),
    spaces: indexBy(serialized.spaces, s => s.uid),
    theorems: indexBy(serialized.theorems, t => t.uid),
    traits: indexBy(serialized.traits, traitId),
    version: serialized.version,
  }
}

type FetchOpts = {
  branch: string
  host?: string
  etag?: string
  fetch?: (input: RequestInfo, init?: RequestInit) => Promise<Response>
}

export function bundleUrl({ branch, host = defaultHost }: FetchOpts): string {
  return `${host}/refs/heads/${branch}.json`
}

export async function fetch(
  opts: FetchOpts,
): Promise<{ bundle: Bundle; etag: string } | undefined> {
  const headers = new Headers()
  if (opts.etag) {
    headers.append('If-None-Match', opts.etag)
  }
  const fetch = opts.fetch || window.fetch
  const response = await fetch(bundleUrl(opts), {
    method: 'GET',
    headers,
  })
  if (response.status === 304) {
    return
  }

  const deserialized = deserialize(await response.json())

  return {
    bundle: deserialized,
    etag: response.headers.get('etag') || '',
  }
}

function indexBy<K, V>(collection: V[], key: (value: V) => K): Map<K, V> {
  return new Map(collection.map((value: V) => [key(value), value]))
}
