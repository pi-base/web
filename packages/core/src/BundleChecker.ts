import { Space } from './Space'
import { Bundle } from './Bundle'
import { deduceTraits } from './Logic'

export type BundleCheckResult =
  | {
      kind: 'contradiction'
      contradiction: { properties: string[]; theorems: string[] }
    }
  | { kind: 'bundle'; bundle: Bundle }

export class BundleChecker {
  constructor(private readonly bundle: Bundle) {}

  check(space: Space): BundleCheckResult {
    const result = deduceTraits(
      this.bundle.implications,
      this.bundle.traitMap({ space }),
    )

    switch (result.kind) {
      case 'contradiction':
        return result
      case 'derivations':
        // TODO: consider if / how to add these
        // for the compile case, we don't want them in the final
        // bundle output, but this check is duplicative with the
        // viewer-side derivation loop
        return { kind: 'bundle', bundle: this.bundle }
    }
  }
}
