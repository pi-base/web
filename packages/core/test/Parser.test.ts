import { expect, it } from 'vitest'
import { parser } from '../src/Parser'

async function parse(input: string) {
  const file = await parser().process(input)
  return String(file)
}

it('parses a citation', async () => {
  expect(await parse('{{doi:123}}')).toEqual(
    '<external-link kind="doi" id="123"></external-link>',
  )
})

it('parses an internal link', async () => {
  expect(await parse('{S123}')).toEqual(
    '<internal-link kind="space" id="123"></internal-link>',
  )
})

it('parses an internal link', async () => {
  expect(await parse('{S123}')).toEqual(
    '<internal-link kind="space" id="123"></internal-link>',
  )
})

it('parses a complex example', async () => {
  const example = `Inline math $2 + 2 = 4$ and display math $$2 + 2 = 4$$.

This is a list of links

* {S000123}
* {P000123}
* {T000123}
* {{doi:123}}
* {{mr:123}}
* {{wikipedia:123}}
* {{mathse:123}}
* {{mo:123}}`

  expect(await parse(example)).toMatchSnapshot()
})
