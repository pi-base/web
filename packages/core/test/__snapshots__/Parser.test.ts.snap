// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parses a complex example 1`] = `
"<p>Inline math <span class=\\"math math-inline\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><mn>2</mn><mo>+</mo><mn>2</mn><mo>=</mo><mn>4</mn></mrow><annotation encoding=\\"application/x-tex\\">2 + 2 = 4</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.7278em;vertical-align:-0.0833em;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2222em;\\"></span><span class=\\"mbin\\">+</span><span class=\\"mspace\\" style=\\"margin-right:0.2222em;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.6444em;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2778em;\\"></span><span class=\\"mrel\\">=</span><span class=\\"mspace\\" style=\\"margin-right:0.2778em;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.6444em;\\"></span><span class=\\"mord\\">4</span></span></span></span></span> and display math <span class=\\"math math-inline\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><mn>2</mn><mo>+</mo><mn>2</mn><mo>=</mo><mn>4</mn></mrow><annotation encoding=\\"application/x-tex\\">2 + 2 = 4</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.7278em;vertical-align:-0.0833em;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2222em;\\"></span><span class=\\"mbin\\">+</span><span class=\\"mspace\\" style=\\"margin-right:0.2222em;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.6444em;\\"></span><span class=\\"mord\\">2</span><span class=\\"mspace\\" style=\\"margin-right:0.2778em;\\"></span><span class=\\"mrel\\">=</span><span class=\\"mspace\\" style=\\"margin-right:0.2778em;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.6444em;\\"></span><span class=\\"mord\\">4</span></span></span></span></span>.</p>
<p>This is a list of links</p>
<ul>
<li><a href=\\"S://000123\\" title=\\"000123\\" class=\\"internal-link\\">000123</a></li>
<li><a href=\\"P://000123\\" title=\\"000123\\" class=\\"internal-link\\">000123</a></li>
<li><a href=\\"T://000123\\" title=\\"000123\\" class=\\"internal-link\\">000123</a></li>
<li><a href=\\"doi://123\\" title=\\"123\\" class=\\"external-link\\">123</a></li>
<li><a href=\\"mr://123\\" title=\\"123\\" class=\\"external-link\\">123</a></li>
<li><a href=\\"wikipedia://123\\" title=\\"123\\" class=\\"external-link\\">123</a></li>
<li><a href=\\"mathse://123\\" title=\\"123\\" class=\\"external-link\\">123</a></li>
<li><a href=\\"mo://123\\" title=\\"123\\" class=\\"external-link\\">123</a></li>
<li><a href=\\"zbmath://123\\" title=\\"123\\" class=\\"external-link\\">123</a></li>
</ul>"
`;

exports[`truncate > preserves math 1`] = `"this is a fairly long sentence and it needs to be summarized because it is too long because it <span class=\\"math math-inline\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><msub><mi>T</mi><mrow><mn>3</mn><mfrac><mn>1</mn><mn>2</mn></mfrac></mrow></msub></mrow><annotation encoding=\\"application/x-tex\\">T_{3\\\\frac{1}{2}}</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:1.1704em;vertical-align:-0.487em;\\"></span><span class=\\"mord\\"><span class=\\"mord mathnormal\\" style=\\"margin-right:0.13889em;\\">T</span><span class=\\"msupsub\\"><span class=\\"vlist-t vlist-t2\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.3448em;\\"><span style=\\"top:-2.7538em;margin-left:-0.1389em;margin-right:0.05em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"sizing reset-size6 size3 mtight\\"><span class=\\"mord mtight\\"><span class=\\"mord mtight\\">3</span><span class=\\"mord mtight\\"><span class=\\"mopen nulldelimiter sizing reset-size3 size6\\"></span><span class=\\"mfrac\\"><span class=\\"vlist-t vlist-t2\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.8443em;\\"><span style=\\"top:-2.656em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"sizing reset-size3 size1 mtight\\"><span class=\\"mord mtight\\"><span class=\\"mord mtight\\">2</span></span></span></span><span style=\\"top:-3.2255em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"frac-line mtight\\" style=\\"border-bottom-width:0.049em;\\"></span></span><span style=\\"top:-3.384em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"sizing reset-size3 size1 mtight\\"><span class=\\"mord mtight\\"><span class=\\"mord mtight\\">1</span></span></span></span></span><span class=\\"vlist-s\\">​…</span></span></span></span></span></span></span></span></span></span></span></span></span></span></span></span></span>"`;
